components/ai_chat/core/common/proto_conversion.cc

```diff
@@ -0,0 +1,149 @@
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/ai_chat/core/common/proto_conversion.h"
+
+#include <utility>
+#include <vector>
+
+#include "base/json/json_reader.h"
+#include "base/logging.h"
+#include "base/values.h"
+#include "url/gurl.h"
+
+namespace ai_chat {
+
+mojom::WebSourcesEventPtr DeserializeWebSourcesEvent(
+    const store::WebSourcesEventProto& proto_event) {
+  auto mojom_event = mojom::WebSourcesEvent::New();
+  mojom_event->sources.reserve(proto_event.sources_size());
+
+  for (const auto& proto_source : proto_event.sources()) {
+    auto mojom_source = mojom::WebSource::New();
+    mojom_source->title = proto_source.title();
+    mojom_source->url = GURL(proto_source.url());
+    if (!mojom_source->url.is_valid()) {
+      DLOG(ERROR) << "Invalid WebSourcesEvent found in database with url: "
+                  << proto_source.url();
+      continue;
+    }
+    mojom_source->favicon_url = GURL(proto_source.favicon_url());
+    if (!mojom_source->favicon_url.is_valid()) {
+      DLOG(ERROR)
+          << "Invalid WebSourcesEvent found in database with favicon url: "
+          << proto_source.favicon_url();
+      continue;
+    }
+    mojom_event->sources.push_back(std::move(mojom_source));
+  }
+  return mojom_event;
+}
+
+void SerializeWebSourcesEvent(const mojom::WebSourcesEventPtr& mojom_event,
+                              store::WebSourcesEventProto* proto_event) {
+  proto_event->clear_sources();
+
+  for (const auto& mojom_source : mojom_event->sources) {
+    if (!mojom_source->url.is_valid() ||
+        !mojom_source->favicon_url.is_valid()) {
+      DLOG(ERROR) << "Invalid WebSourcesEvent found for persistence, with url: "
+                  << mojom_source->url.spec()
+                  << " and favicon url: " << mojom_source->favicon_url.spec();
+      continue;
+    }
+    store::WebSourceProto* proto_source = proto_event->add_sources();
+    proto_source->set_title(mojom_source->title);
+    proto_source->set_url(mojom_source->url.spec());
+    proto_source->set_favicon_url(mojom_source->favicon_url.spec());
+  }
+}
+
+mojom::ToolUseEventPtr DeserializeToolUseEvent(
+    const store::ToolUseEventProto& proto_event) {
+  auto mojom_event = mojom::ToolUseEvent::New(
+      proto_event.tool_name(), proto_event.id(), proto_event.arguments_json(),
+      std::nullopt, std::nullopt);
+
+  // Convert arguments from JSON string to mojo DictionaryValue
+  if (!proto_event.arguments_json().empty()) {
+    // We assume stored arguments is safe JSON and we don't need to parse
+    // in a separate process.
+    auto parsed_arguments =
+        base::JSONReader::ReadDict(proto_event.arguments_json());
+    if (parsed_arguments) {
+      mojom_event->arguments = std::move(parsed_arguments.value());
+    }
+  } else {
+    mojom_event->arguments = base::Value::Dict();
+  }
+
+  // Convert output ContentBlocks
+  if (proto_event.output_size() > 0) {
```

Comment:
I don't think we need this `if` block - we loop over the items, and if there are no items we won't do anything.